src/System.Management.Automation/engine/parser/tokenizer.cs

```diff
@@ -3268,13 +3278,33 @@ private static bool TryGetNumberValue(string strNum, bool hex, bool real, char s
                             {
                                 d = -0.0;
                             }
-                            if (suffix == 'l' || suffix == 'L')
+
+                            if (suffix == NumberSuffixFlags.Long)
                             {
-                                result = ((long)Convert.ChangeType(d, typeof(long), CultureInfo.InvariantCulture)) * multiplier;
+                                result = ((long)Convert.ChangeType(d, typeof(long), CultureInfo.InvariantCulture) * multiplier);
+                            }
+                            else if (suffix == NumberSuffixFlags.Short)
+                            {
+                                result = ((short)Convert.ChangeType(d, typeof(short), CultureInfo.InvariantCulture) * (short)multiplier);
+                            }
+                            else if (suffix.HasFlag(NumberSuffixFlags.Unsigned))
+                            {
+                                if (suffix.HasFlag(NumberSuffixFlags.Long))
+                                {
+                                    result = ((ulong)Convert.ChangeType(d, typeof(ulong), CultureInfo.InvariantCulture) * (ulong)multiplier);
+                                }
+                                else if (suffix.HasFlag(NumberSuffixFlags.Short))
+                                {
+                                    result = ((ushort)Convert.ChangeType(d, typeof(ushort), CultureInfo.InvariantCulture) * (ushort)multiplier);
+                                }
+                                else
+                                {
+                                    result = ((uint)Convert.ChangeType(d, typeof(uint), CultureInfo.InvariantCulture) * (uint)multiplier);
+                                }
                             }
                             else
                             {
-                                result = d * multiplier;
+                                result = d;
```

Comment:
Ah, I thought I got all those. Correcting...

(I was trying to see if I could do the multiplier before the conversions to avoid throwing too much here, but it doesn't work *because* Int64.MaxValue is too high to retain precision with Double values in the upper range there.)